feat: allow global host commands to be run from anywhere, for #4416

GuySartorelli opened the pull request:
<!-- 
  PR titles have very precise rules, please read 
  https://ddev.readthedocs.io/en/stable/developers/building-contributing/#open-pull-requests
-->

## The Issue

Custom global host commands currently only be run inside a project directory - even if they don't need a project in order to do their stuff.

A great example of this is `ddev self-upgrade` which should be able to run from anywhere - but currently it has to be run from inside a project directory.

## How This PR Solves The Issue

This PR adds a new `CanRunGlobally` annotation. If custom global host commands set this annotation to `true`, they will be added to the cobra command set, and can be run from anywhere, even outside project directories.

## Manual Testing Instructions

1. Run `ddev help` - you should see the `self-upgrade` command there.
1. Run `ddev self-upgrade` - it should work from anywhere, even outside a project directory.
1. Add `## CanRunGlobally: true` to a custom global host command, then perform the first two steps again, but with your custom command rather than `self-upgrade`.

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

I've added checks for the following the following:
- Check a global host command with `CanRunGlobally: true` shows and can be run from arbitrary working directories
- Check a global host command with `CanRunGlobally: true` still shows and can be run from inside a project directory
- Check commands without `CanRunGlobally: true` do not show from arbitrary working directories

This is my first time really looking at how to add tests and I've gone in with a fairly light touch - if you want more test coverage I'll be happy to add it.

## Related Issue Link(s)

- https://github.com/ddev/ddev/issues/4416

**NOTE:** This PR will be one of several iterative features that together will fulfil the requirements set forth in that issue. The issue should not be closed until all of its requirements are fulfilled.

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->
Anyone who maintains a DDEV extension or who has some custom commands in projects they manage should review their custom global host commands, and add the new `CanRunGlobally` annotation to any commands that should be allowed to run even outside of project directories.

GuySartorelli commented on cmd/ddev/cmd/commands.go:
```diff
@@ -35,8 +36,30 @@ func IsUserDefinedCustomCommand(cmd *cobra.Command) bool {
 // and if it finds them adds them to Cobra's commands.
 func addCustomCommands(rootCmd *cobra.Command) error {
+	// Custom commands are shell scripts - so we can't use them on windows without bash.
+	if runtime.GOOS == "windows" {
+		windowsBashPath := util.FindBashPath()
+		if windowsBashPath == "" {
+			fmt.Println("Unable to find bash.exe in PATH, not loading custom commands")
+			return nil
+		}
+	}
```
This was further down, inside a loop for some reason - but it just needs to be run once and should be run early to avoid unnecessary processing

GuySartorelli commented on cmd/ddev/cmd/commands.go:
```diff
@@ -266,18 +298,23 @@ func makeHostCmd(app *ddevapp.DdevApp, fullPath, name string) func(*cobra.Comman
+			status, _ := app.SiteStatus()
+			app.DockerEnv()
+			_ = os.Setenv("DDEV_PROJECT_STATUS", status)
+		} else {
+			_ = os.Setenv("DDEV_PROJECT_STATUS", "")
+		}
```
Not sure if we need this or not? I figured it's safer to have it and not need it, rather than not have it and find out later we do.

stasadev commented:
Wow!

> The rules for PR titles say that I must use fixes #4416 in the PR title because an issue exists

You can use `for #4416` instead of `fixes #4416` .

> Docs CI failure seems to be for an unrelated broken link

Restarted failed jobs.

GuySartorelli commented:
I've added some tests and updated the "Automated Testing Overview" section of the PR description.
Let me know if you want anything changed, more test coverage, etc.

GuySartorelli commented:
I tried to test this on gitpod using https://gitpod.io/#https://github.com/ddev/ddev/pulls/5702 but when I ran `ddev -v` it showed that it was actually running on commit `9bd9054` (i.e. the latest commit on the `master` branch)... so I'm not sure if I was doing something wrong there, or if something is misconfigured, or something?

Anyway, that was more trying that out for fun - I did run this locally and it worked as expected.

rfay commented:
I just opened on gitpod and got
```
$ ddev --version
ddev version feaedb50
```

which is correct. I used the "open" button provided by the gitpod app, 
<img width="166" alt="image" src="https://github.com/ddev/ddev/assets/112444/34b04e44-143e-4f3c-8189-3e4a62053fb8">

It links to `https://gitpod.io/?autostart=true#https://github.com/ddev/ddev/pull/5702` which should be the same as your result. 

Would be interested to know what might have gone wrong for you. Gitpod is a great resource for these things. 

I am interested to note that gitpod no longer adds a comment with the link. That was a nice feature. 

I tried it with your link, `https://gitpod.io/#https://github.com/ddev/ddev/pulls/5702` and sure enough, the code in /workspace/ddev was upstream/master instead of the PR. 

Mine `https://gitpod.io/?autostart=true#https://github.com/ddev/ddev/pull/5702` checks out the branch properly:
```
gitpod /workspace/ddev (20240116_GuySartorelli_truly-global-commands) $ git status
On branch 20240116_GuySartorelli_truly-global-commands
Your branch is up to date with 'origin/20240116_GuySartorelli_truly-global-commands'.
```

Yours,  `https://gitpod.io/#https://github.com/ddev/ddev/pulls/5702`, 
```
gitpod /workspace/ddev (master) $ git status
On branch master
Your branch is up to date with 'origin/master'.
```

I have to assume this is a new bug in gitpod? Or perhaps it's just 
* https://github.com/gitpod-io/gitpod/issues/13991

which never actually got fixed?

rfay commented:
Ah, the problem is your URL had `pulls` in it, not `pull`.  (And that's because the doc says `pulls`). (It works fine with `pull`). I'll do docs PR.

GuySartorelli commented:
I've renamed the `RunsWithoutProject` annotation to `CanRunGlobally`.

This is a more immediately clear and better descriptive name for what this annotation does, and leaves its usage open to the behaviour described in https://github.com/ddev/ddev/issues/4416#issuecomment-1901847995 in the future.

GuySartorelli commented:
Build kits ci failure looks unrelated

rfay commented:
I invited you to buildkite so you can restart tests (and more easily examine them), also to ddev team on GitHub. I think that lets you restart GitHub tests.

GuySartorelli commented:
Awesome, thank you! 💙

GuySartorelli commented:
This should be good to go

stasadev commented:
I will review this later this week.

stasadev reviewed (approved):
This is a great addition to DDEV!

Works as expected, thank you @GuySartorelli.
